paddle/memory/detail/system_allocator.cc

```diff
@@ -77,45 +78,20 @@ void* GPUAllocator::Alloc(size_t& index, size_t size) {
   // CUDA documentation doesn't explain if cudaMalloc returns nullptr
   // if size is 0.  We just make sure it does.
   if (size <= 0) return nullptr;
-
-  size_t available = 0;
-  size_t capacity = 0;
-  paddle::platform::GpuMemoryUsage(available, capacity);
-
-  // Reserve memory for page tables, etc.
-  size_t reserving = 0.05 * capacity + paddle::platform::GpuMinChunkSize();
-  size_t usable = available > reserving ? available - reserving : 0;
-
-  // If remaining size no less than expected size, using general
-  // cudaMalloc to allocate GPU memory.
-  void* p = 0;
-  if (size <= usable) {
-    cudaError_t result = cudaMalloc(&p, size);
-    if (result == cudaSuccess) {
-      index = 0;
-      gpu_alloc_size_ += size;
-      return p;
-    }
-  }
-
-  // If remaining size less than expected size or cudaMalloc failed,
-  // cudaMallocHost will be considered as a fallback allocator.
-  //
-  // NOTE: here, we use GpuMaxAllocSize() as the maximum memory size
-  // of host fallback allocation. Allocates too much would reduce
-  // the amount of memory available to the underlying system for paging.
-  usable = paddle::platform::GpuMaxAllocSize() - fallback_alloc_size_;
-
-  if (size > usable) return nullptr;
-
-  cudaError_t result = cudaMallocHost(&p, size);
```

Comment:
Why remove `cudaMallocHost`?

Most of the memory required for training deep recurrent networks is used to store activations through each layer for use by back propagation, not to store the parameters of the network. For example, storing the weights for a 70M parameter network with 9 layers requires approximately 280 MB of memory, but storing the activations for a batch of 64, seven-second utterances requires 1.5 GB of memory. TitanX GPUs include 12GB of GDDR5 RAM, and sometimes very deep networks can exceed the GPU memory capacity when processing long utterances. This can happen unpredictably,  it is desirable to avoid a catastrophic failure when this occurs. 

The combination of fast memory allocation with a fallback mechanism that allows us to slightly overflow available GPU memory in exceptional cases makes the system significantly simpler, more robust, and more efficient. 

This memory can be accessed directly by the GPU by forwarding individual memory transactions over PCIe at reduced bandwidth, and it allows a model to continue to make progress even after encountering an outlier.

@reyoung @chengduoZH @wangkuiyi

Reply:
Deep Speech 2 section 4.3, https://arxiv.org/pdf/1512.02595.pdf

Reply:
From `Zen of python`

```
Explicit is better than implicit.
Simple is better than complex.
```

`cudaMallocHost` will be invoked IMPLICITLY when out of memory in the previous implementation. The performance is very poor when uses `cudaMallocHost` to allocate memory and run the kernel on GPU and since it is an implicit behaviour, it is hard to debug before.

It may be better to fail fast when out of memory. It is explicit and simple.

If this feature is needed, we can also implement another decorator rather than combine these logics together. For example:

```C++
class CUDAFallbackAllocator {
public:
  void* alloc(size_t size) {
    void* ptr = allocator_->alloc(size);
    if (ptr == nullptr) {
        return cudaMallocHost(size);
    } else {
       return ptr;
    }
  }

private:
  CUDASystemAllocator* allocator_;
}
```

Reply:
I see, thanks.